container/sagemaker/python_service.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import json
+import os
+
+import importlib.util
+import logging
+import re
+from collections import namedtuple
+
+import falcon
+import requests
+
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger(__name__)
+
+DEFAULT_CONTENT_TYPE = 'application/json'
+DEFAULT_ACCEPT_HEADER = 'application/json'
+CUSTOM_ATTRIBUTES = 'X-Amzn-SageMaker-Custom-Attributes'
+
+Context = namedtuple('Context',
+                     'model_name, model_version, method, rest_uri, grpc_uri, '
+                     'custom_attributes, request_content_type, accept_header')
+
+
+class InvocationResource(object):
+
+    def __init__(self):
+        if 'SAGEMAKER_SAFE_PORT_RANGE' in os.environ:
+            port_range = os.environ['SAGEMAKER_SAFE_PORT_RANGE']
+            parts = port_range.split('-')
+            low = int(parts[0])
+            high = int(parts[1])
+            if low + 2 > high:
+                raise ValueError('not enough ports available in SAGEMAKER_SAFE_PORT_RANGE ({})'
+                                 .format(port_range))
+            self._tfs_grpc_port = str(low)
+            self._tfs_rest_port = str(low + 1)
+        else:
+            self._tfs_grpc_port = '9000'
+            self._tfs_rest_port = '8501'
+
+        self.handler = None
+        self.input_handler = None
+        self.output_handler = None
+        self._import_handlers()
+
+    def on_post(self, req, res):
+        context, data = self._parse_request(req)
+
+        if self.handler:
+            res.status = falcon.HTTP_200
+            res.body = self.handler(data, context)
+        elif self.input_handler and self.output_handler:
+            processed_input = self.input_handler(data, context)
+            response = requests.post(context.rest_uri, data=json.dumps(processed_input))
+            prediction, content_type = self.output_handler(response, context)
```

Comment:
we are assuming `output_handler` returns a python object that can be dumped to json, but we need to support arbitrary return types.

our options here are to require return type bytes (which we'd just pass to resp.data), or allow bytes, string, other and pass the value to resp.data, resp.body, resp.media depending on the type. the flexible approach is better, but we'd need to update the various doc comments to match. for the other/media option, we should describe that as a dict -> json handler rather than identifying and falcon-specific features.

see falcon docs on  [media](https://falcon.readthedocs.io/en/stable/api/request_and_response.html#falcon.Response.media),  [body](https://falcon.readthedocs.io/en/stable/api/request_and_response.html#falcon.Response.body),  and [data](https://falcon.readthedocs.io/en/stable/api/request_and_response.html#falcon.Response.data) for more context